sdk/core/Azure.Core/src/ResourceIdentifier.cs

```diff
@@ -39,137 +64,178 @@ public sealed class ResourceIdentifier : IEquatable<ResourceIdentifier>, ICompar
+            ReadOnlySpan<char> thirdWord = PopNextWord(ref remaining);
+            if (firstWord.Equals(ProvidersKey.AsSpan(), StringComparison.OrdinalIgnoreCase))
             {
-                if (partsCount == 2 || parts[startIndex + 2].Equals(ProvidersKey, StringComparison.OrdinalIgnoreCase))
+                if (thirdWord.IsEmpty || thirdWord.Equals(ProvidersKey.AsSpan(), StringComparison.OrdinalIgnoreCase))
                 {
                     //provider resource can only be on a tenant or a subscription parent
-                    if (parent.ResourceType == SubscriptionResourceType || parent.ResourceType == TenantResourceType)
-                        return AppendNext(new ResourceIdentifier(parent, ProviderResourceType, parts[startIndex + 1], true), parts, startIndex + 2, length);
+                    if (parent.ResourceType == ResourceType.Subscription || parent.ResourceType == ResourceType.Tenant)
+                    {
+                        nextWord = thirdWord;
+                        return new ResourceIdentifierParts(parent, ResourceType.Provider, secondWord.ToString(), true, SpecialType.Provider);
+                    }
 
-                    throw new ArgumentOutOfRangeException(nameof(parts), $"Provider resource can only come after the root or {SubscriptionsKey}.");
+                    throw new ArgumentOutOfRangeException("resourceId", $"Provider resource can only come after the root or {SubscriptionsKey}.");
                 }
 
-                if (partsCount > 3)
-                    return AppendNext(new ResourceIdentifier(parent, parts[startIndex + 1], parts[startIndex + 2], parts[startIndex + 3]), parts, startIndex + 4, length);
+                ReadOnlySpan<char> fourthWord = PopNextWord(ref remaining);
+                if (!fourthWord.IsEmpty)
+                {
+                    nextWord = PopNextWord(ref remaining);
+                    SpecialType? specialType = thirdWord.Equals(LocationsKey.AsSpan(), StringComparison.OrdinalIgnoreCase) ? SpecialType.Location : null;
+                    return new ResourceIdentifierParts(parent, new ResourceType(secondWord.ToString(), thirdWord.ToString()), fourthWord.ToString(), true, specialType);
+                }
             }
             else
             {
-                return AppendNext(new ResourceIdentifier(parent, firstElement, parts[startIndex + 1]), parts, startIndex + 2, length);
+                nextWord = thirdWord;
+                return new ResourceIdentifierParts(parent, ChooseResourceType(firstWord, parent, out SpecialType? specialType), secondWord.ToString(), false, specialType);
             }
 
-            throw new ArgumentOutOfRangeException(nameof(parts), "Invalid resource id.");
+            throw new ArgumentOutOfRangeException("resourceId", "Invalid resource id.");
+        }
+#pragma warning restore CA2208 // Instantiate argument exceptions correctly
```

Comment:
Does this mean that the parameters are backwards? I think we should fix these and not disable the warning.

Reply:
Ah -- so this is about the fact that we are using 'resourceid' as the parameter name?

Reply:
I guess my point was what do we think is more useful to a customer who called a ctor with 1 param named `resourceId`

An error that says `parts` was invalid
An error that says `resourceId` was invalid

Reply:
I was just wondering if there's a way to restructure the code or rename parameters so we can throw the right exception from a customer perspective and not have to disable the warning.

If you decide to keep disabling the warning, please do scope it to the single lines where it applies.

Reply:
After changing to lazy initialization there is no point in throwing ArgumentException from the property access. I switched to FormatException, other "parsing" types like `Uri` use it in a similar situation.